mauth_client/middlewares/asgi.py

```diff
@@ -0,0 +1,115 @@
+        headers = {decode(k): decode(v) for k, v in scope["headers"]}
+
+        events, body = await self._get_body(receive)
+
+        signable = RequestSignable(
+            method=scope["method"],
+            url=url,
+            body=body,
+        )
+        signed = Signed.from_headers(headers)
+        authenticator = LocalAuthenticator(signable, signed, logger)
+        is_authentic, status, message = authenticator.is_authentic()
+
+        if is_authentic:
+            # asgi spec calls for passing a copy of the scope rather than mutating it
+            # note: deepcopy will blow up with infi recursion due to objects in some values
+            scope_copy = scope.copy()
+            scope_copy[ENV_APP_UUID] = signed.app_uuid
+            scope_copy[ENV_AUTHENTIC] = True
+            scope_copy[ENV_PROTOCOL_VERSION] = signed.protocol_version()
+            await self.app(scope_copy, self._fake_receive(events), send)
+        else:
+            await self._send_response(send, status, message)
+
+    def _validate_configs(self) -> None:
+        # Validate the client settings (APP_UUID, PRIVATE_KEY)
+        if not all([Config.APP_UUID, Config.PRIVATE_KEY]):
+            raise TypeError("MAuthASGIMiddleware requires APP_UUID and PRIVATE_KEY")
+        # Validate the mauth settings (MAUTH_BASE_URL, MAUTH_API_VERSION)
+        if not all([Config.MAUTH_URL, Config.MAUTH_API_VERSION]):
+            raise TypeError("MAuthASGIMiddleware requires MAUTH_URL and MAUTH_API_VERSION")
+
+    async def _get_body(
+        self, receive: ASGIReceiveCallable
+    ) -> Tuple[List[ASGIReceiveEvent], str]:
+        body = b""
+        more_body = True
+        events = []
+
+        while more_body:
+            event = await receive()
+            body += event.get("body", b"")
+            more_body = event.get("more_body", False)
+            events.append(event)
+        return (events, decode(body))
+
+    async def _send_response(self, send: ASGISendCallable, status: int, msg: str) -> None:
+        await send({
+            "type": "http.response.start",
+            "status": status,
+            "headers": [(b"content-type", b"application/json")],
+        })
+        body = {"errors": {"mauth": [msg]}}
+        await send({
+            "type": "http.response.body",
+            "body": json.dumps(body).encode("utf-8"),
+        })
+
+    def _fake_receive(self, events: List[ASGIReceiveEvent]) -> ASGIReceiveCallable:
+        """
+        Create a fake, async receive function using an iterator of the events
+        we've already read. This will be passed to downstream middlewares/apps
+        instead of the usual receive fn, so that they can also "receive" the
+        body events.
+        """
+        events_iter = iter(events)
+
+        async def _receive() -> ASGIReceiveEvent:
+            try:
+                return next(events_iter)
+            except StopIteration:
+                pass
+        return _receive
```

Comment:
@danielloganking @ykitamura-mdsol ok, i think this is best/only solution.  as we are reading the events and building the body using the given ASGI `receive` function, we also save those events.  we wrap the saved events in an iterator and fake function to pass to downstream middlewares/apps instead of the real `receive`.  then when they read from our fake function, they will receive the events in the same order/fashion.

as i mentioned before, this is pretty much defeats one of the main advantages of async: being able to process the body as a stream.  however, we don't really have a choice as, with our authentication scheme, we need to have the body to be able to verify the signature.  additionally, we wouldn't want to start processing pieces of the body as a stream if the request turned out to be inauthentic.

i don't think we have to worry about upstream middlewares stealing the body before us because:

1. it is kinda against the whole async idea as i mentioned above.
2. if they do it, we should expect them to do similar to what we are doing here to as to not break downstream.

Reply:
theoretically we could also return the entire body in one event like:

```python
{
    "type":  "http.request",
    "body": b"the whole body",
    "more_body": False,
}
```

but it seems more correct to me to just give the exact same events that we got.

docs for the receive events: https://asgi.readthedocs.io/en/latest/specs/www.html#request-receive-event

Reply:
I have zero experience with ASGI so I might not be in the right position to decide, but it looks good.
It seems like consuming body in middleware is a pain point in ASGI...
- https://github.com/encode/starlette/issues/495

Reply:
This makes my head hurt but I agree it seems to resolve the early body exhaustion issue.

Reply:
yes, i saw that github issue as well.  i guess maybe it would make more sense if they exposed the stream somehow.. but then again, the way the spec is now kinda ensures performance by preventing multiple middlewares from re-reading the same stream.